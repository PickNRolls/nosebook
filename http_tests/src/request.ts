import supertest, { Test } from 'supertest';
import TestAgent from 'supertest/lib/agent';

const HOST = 'http://backend:8080';
const SESSION_HEADER_KEY = 'X-Auth-Session-Id';

class TestAgentWrapper {
  private host: string;
  private agent: TestAgent;

  public constructor(host: string) {
    this.host = host;
    this.agent = supertest(host);
  }

  public unwrap(): TestAgent {
    return this.agent;
  }

  public extend(prefixUrl: string): TestAgentWrapper {
    return new TestAgentWrapper(this.host + prefixUrl);
  }

  private hook<M extends 'get' | 'post'>(method: M, url: string): Test {
    const test = this.agent[method](url);

    test.set(SESSION_HEADER_KEY, 'bb23af03-be50-4bce-b729-b259b2e02e54');

    return test;
  }

  public get(url: string): Test {
    return this.hook('get', url);
  }

  public post(url: string): Test {
    return this.hook('post', url);
  }
};

export default new TestAgentWrapper(HOST);


