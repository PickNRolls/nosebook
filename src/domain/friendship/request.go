package domainfriendship

import (
	"nosebook/src/errors"
	"time"

	"github.com/google/uuid"
)

type FriendRequest struct {
	RequesterId uuid.UUID `json:"requesterId" db:"requester_id"`
	ResponderId uuid.UUID `json:"responderId" db:"responder_id"`
	Message     string    `json:"message" db:"message"`
	Accepted    bool      `json:"accepted" db:"accepted"`
	Viewed      bool      `json:"viewed" db:"viewed"`
	CreatedAt   time.Time `json:"createdAt" db:"created_at"`

	events []Event
}

func New(
	requesterId uuid.UUID,
	responderId uuid.UUID,
	message string,
	accepted bool,
	viewed bool,
	createdAt time.Time,
	raiseCreatedEvent bool,
) *FriendRequest {
	output := &FriendRequest{
		RequesterId: requesterId,
		ResponderId: responderId,
		Message:     message,
		Accepted:    accepted,
		Viewed:      viewed,
		CreatedAt:   createdAt,
		events:      make([]Event, 0),
	}

	if raiseCreatedEvent {
		output.raiseEvent(NewCreatedEvent())
	}

	return output
}

func (this *FriendRequest) raiseEvent(event Event) {
	this.events = append(this.events, event)
}

func (this *FriendRequest) Events() []Event {
	return this.events
}

func (this *FriendRequest) ViewBy(userId uuid.UUID) *errors.Error {
	if userId != this.ResponderId {
		return errors.New("Friendship Error", "Только получатель может отметить заявку просмотренной")
	}

	this.Viewed = true
	this.raiseEvent(NewViewedEvent())

	return nil
}

func (this *FriendRequest) AcceptBy(userId uuid.UUID) *errors.Error {
	if userId != this.ResponderId {
		return errors.New("Friendship Error", "Только получатель может принять заявку")
	}

	this.Accepted = true
	this.Viewed = true
	this.raiseEvent(NewAcceptedEvent())

	return nil
}

func (this *FriendRequest) DenyBy(userId uuid.UUID) *errors.Error {
	if userId != this.ResponderId {
		return errors.New("Friendship Error", "Только получатель может отклонить заявку")
	}

	this.Accepted = false
	this.Viewed = true
	this.raiseEvent(NewDeniedEvent())

	return nil
}

func (this *FriendRequest) RemoveBy(userId uuid.UUID) *errors.Error {
	if !this.Accepted {
		return errors.New("Friendship Error", "Нельзя удалить непринятую заявку")
	}

	this.Accepted = false
	previousRequesterId := this.RequesterId
	previousResponderId := this.ResponderId

	if this.RequesterId == userId {
		this.RequesterId, this.ResponderId = this.ResponderId, this.RequesterId
	}

	this.raiseEvent(NewRemovedEvent(previousRequesterId, previousResponderId))

	return nil
}
