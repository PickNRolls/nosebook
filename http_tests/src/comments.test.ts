import supertest from 'supertest';
import request from './request';

describe('comments', () => {
  let comments = request.extend({ prefixUrl: '/comments' });
  let createdCommentId: string;

  test('GET /', async () => {
    const postId = 'c7b7bf17-38f9-4ed5-b0a8-501a90f7c8e7';

    let lastResponse = await comments
      .get('/')
      .query({
        postId,
      })
      .expect(200)

    expect(lastResponse.body).toMatchSnapshot();

    lastResponse = await comments
      .get('/')
      .query({
        postId,
        next: lastResponse.body.data.next,
      })
      .expect(200);

    expect(lastResponse.body).toMatchSnapshot();
  });

  test('POST /publish-on-post', async () => {
    let response = await comments
      .post('/publish-on-post')
      .send({
        id: 'c7b7bf17-38f9-4ed5-b0a8-501a90f7c8e7',
        message: 'my test comment',
      })
      .expect(200);

    expect(response.body).toStrictEqual(expect.objectContaining({
      data: {
        id: expect.any(String),
      },
      ok: true,
    }));
    createdCommentId = response.body.data.id;
  });

  test('POST /remove', async () => {
    const response = await comments
      .post('/remove')
      .send({
        id: createdCommentId,
      }).expect(200);

    expect(response.body).toStrictEqual(expect.objectContaining({
      data: {
        id: createdCommentId,
      },
      ok: true,
    }));
  });

  describe('permissions', () => {
    test('GET /:id, has permissions', async () => {
      let response = await comments
        .get(`/120c79b8-3927-48b7-a308-1ffd3db6036f`)
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });

    test('GET /:id, has no permissions', async () => {
      let response = await comments
        .get('/d0023f4d-8d7f-4907-9438-d2ed2a9661f0')
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });
  });
});

