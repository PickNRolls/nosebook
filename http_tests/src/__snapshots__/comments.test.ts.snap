// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`comments GET / 1`] = `
{
  "data": {
    "data": [
      {
        "author": {
          "firstName": "Test",
          "id": "ed1a3fd0-4d0b-4961-b4cd-cf212357740d",
          "lastName": "Tester",
          "lastOnlineAt": "2024-08-10T10:10:10Z",
          "nick": "test_tester",
          "online": true,
        },
        "createdAt": "2024-02-16T15:36:38Z",
        "id": "120c79b8-3927-48b7-a308-1ffd3db6036f",
        "likes": {
          "count": 0,
          "liked": false,
          "randomFiveLikers": [],
        },
        "message": "comment for permissions test",
        "permissions": {
          "remove": true,
          "update": true,
        },
      },
      {
        "author": {
          "firstName": "Ilya",
          "id": "1ae02f69-ea1a-4308-b825-0e5896e652e4",
          "lastName": "Blinkov",
          "lastOnlineAt": "2024-08-10T10:01:02Z",
          "nick": "drugtester",
          "online": false,
        },
        "createdAt": "2024-02-16T15:36:55Z",
        "id": "d0023f4d-8d7f-4907-9438-d2ed2a9661f0",
        "likes": {
          "count": 0,
          "liked": false,
          "randomFiveLikers": [],
        },
        "message": "3nd comment message",
        "permissions": {
          "remove": false,
          "update": false,
        },
      },
      {
        "author": {
          "firstName": "Ilya",
          "id": "1ae02f69-ea1a-4308-b825-0e5896e652e4",
          "lastName": "Blinkov",
          "lastOnlineAt": "2024-08-10T10:01:02Z",
          "nick": "drugtester",
          "online": false,
        },
        "createdAt": "2024-02-16T15:37:55Z",
        "id": "d0023f4d-8d7f-4907-9438-d2ed2a9661f1",
        "likes": {
          "count": 0,
          "liked": false,
          "randomFiveLikers": [],
        },
        "message": "4nd comment message",
        "permissions": {
          "remove": false,
          "update": false,
        },
      },
      {
        "author": {
          "firstName": "Ilya",
          "id": "1ae02f69-ea1a-4308-b825-0e5896e652e4",
          "lastName": "Blinkov",
          "lastOnlineAt": "2024-08-10T10:01:02Z",
          "nick": "drugtester",
          "online": false,
        },
        "createdAt": "2024-02-16T15:38:55Z",
        "id": "d0023f4d-8d7f-4907-9438-d2ed2a9661f2",
        "likes": {
          "count": 0,
          "liked": false,
          "randomFiveLikers": [],
        },
        "message": "5nd comment message",
        "permissions": {
          "remove": false,
          "update": false,
        },
      },
      {
        "author": {
          "firstName": "Ilya",
          "id": "1ae02f69-ea1a-4308-b825-0e5896e652e4",
          "lastName": "Blinkov",
          "lastOnlineAt": "2024-08-10T10:01:02Z",
          "nick": "drugtester",
          "online": false,
        },
        "createdAt": "2024-02-16T15:39:55Z",
        "id": "d0023f4d-8d7f-4907-9438-d2ed2a9661f3",
        "likes": {
          "count": 0,
          "liked": false,
          "randomFiveLikers": [],
        },
        "message": "6nd comment message",
        "permissions": {
          "remove": false,
          "update": false,
        },
      },
    ],
    "next": "d0023f4d-8d7f-4907-9438-d2ed2a9661f3/2024-02-16T15:39:55Z",
    "totalCount": 8,
  },
  "ok": true,
}
`;

exports[`comments GET / 2`] = `
{
  "data": {
    "data": [
      {
        "author": {
          "firstName": "Ilya",
          "id": "1ae02f69-ea1a-4308-b825-0e5896e652e4",
          "lastName": "Blinkov",
          "lastOnlineAt": "2024-08-10T10:01:02Z",
          "nick": "drugtester",
          "online": false,
        },
        "createdAt": "2024-02-16T15:40:55Z",
        "id": "d0023f4d-8d7f-4907-9438-d2ed2a9661f4",
        "likes": {
          "count": 0,
          "liked": false,
          "randomFiveLikers": [],
        },
        "message": "7nd comment message",
        "permissions": {
          "remove": false,
          "update": false,
        },
      },
      {
        "author": {
          "firstName": "Ilya",
          "id": "1ae02f69-ea1a-4308-b825-0e5896e652e4",
          "lastName": "Blinkov",
          "lastOnlineAt": "2024-08-10T10:01:02Z",
          "nick": "drugtester",
          "online": false,
        },
        "createdAt": "2024-02-16T15:41:55Z",
        "id": "d0023f4d-8d7f-4907-9438-d2ed2a9661f5",
        "likes": {
          "count": 0,
          "liked": false,
          "randomFiveLikers": [],
        },
        "message": "8nd comment message",
        "permissions": {
          "remove": false,
          "update": false,
        },
      },
      {
        "author": {
          "firstName": "Ilya",
          "id": "1ae02f69-ea1a-4308-b825-0e5896e652e4",
          "lastName": "Blinkov",
          "lastOnlineAt": "2024-08-10T10:01:02Z",
          "nick": "drugtester",
          "online": false,
        },
        "createdAt": "2024-02-16T15:42:55Z",
        "id": "d0023f4d-8d7f-4907-9438-d2ed2a9661f6",
        "likes": {
          "count": 0,
          "liked": false,
          "randomFiveLikers": [],
        },
        "message": "9nd comment message",
        "permissions": {
          "remove": false,
          "update": false,
        },
      },
    ],
    "totalCount": 8,
  },
  "ok": true,
}
`;

exports[`comments permissions GET /:id, has no permissions 1`] = `
{
  "data": {
    "author": {
      "firstName": "Ilya",
      "id": "1ae02f69-ea1a-4308-b825-0e5896e652e4",
      "lastName": "Blinkov",
      "lastOnlineAt": "2024-08-10T10:01:02Z",
      "nick": "drugtester",
      "online": false,
    },
    "createdAt": "2024-02-16T15:36:55Z",
    "id": "d0023f4d-8d7f-4907-9438-d2ed2a9661f0",
    "likes": {
      "count": 0,
      "liked": false,
      "randomFiveLikers": [],
    },
    "message": "3nd comment message",
    "permissions": {
      "remove": false,
      "update": false,
    },
  },
  "ok": true,
}
`;

exports[`comments permissions GET /:id, has permissions 1`] = `
{
  "data": {
    "author": {
      "firstName": "Test",
      "id": "ed1a3fd0-4d0b-4961-b4cd-cf212357740d",
      "lastName": "Tester",
      "lastOnlineAt": "2024-08-10T10:10:10Z",
      "nick": "test_tester",
      "online": true,
    },
    "createdAt": "2024-02-16T15:36:38Z",
    "id": "120c79b8-3927-48b7-a308-1ffd3db6036f",
    "likes": {
      "count": 0,
      "liked": false,
      "randomFiveLikers": [],
    },
    "message": "comment for permissions test",
    "permissions": {
      "remove": true,
      "update": true,
    },
  },
  "ok": true,
}
`;
